internal/authn/oidc/oidc_test.go

```diff
@@ -0,0 +1,200 @@
+		},
+		{
+			testDescription: "when_token's_issuer_does_not_match_the_one_provided_in_the_server_configuration,_MUST_return_'invalid_issuer'_error",
+			testSetup: func() (*RemoteOidcAuthenticator, context.Context, error) {
+				return quickConfigSetup(
+					"kid_1",
+					"kid_1",
+					"right_issuer",
+					"",
+					jwt.MapClaims{
+						"iss": "wrong_issuer",
+					},
+					nil,
+				)
+			},
+			expectedError: "invalid issuer",
+		},
+		{
+			testDescription: "when_token's_audience_does_not_match_the_one_provided_in_the_server_configuration,_MUST_return_'invalid_audience'_error",
+			testSetup: func() (*RemoteOidcAuthenticator, context.Context, error) {
+				return quickConfigSetup(
+					"kid_1",
+					"kid_1",
+					"right_issuer",
+					"right_audience",
+					jwt.MapClaims{
+						"iss": "right_issuer",
+						"aud": "wrong_audience",
+					},
+					nil,
+				)
+			},
+		},
+		{
+			testDescription: "when_the_subject_of_the_token_is_not_a_string,_MUST_return_'invalid_subject'_error",
+			testSetup: func() (*RemoteOidcAuthenticator, context.Context, error) {
+				return quickConfigSetup(
+					"kid_1",
+					"kid_1",
+					"right_issuer",
+					"right_audience",
+					jwt.MapClaims{
+						"iss": "right_issuer",
+						"aud": "right_audience",
+						"sub": 12,
+					},
+					nil,
+				)
+			},
+			expectedError: "invalid subject",
+		},
+	}
+
+	for _, testC := range testCases {
+		t.Run(testC.testDescription, func(t *testing.T) {
+			oidc, requestContext, err := testC.testSetup()
+			_, err = oidc.Authenticate(requestContext)
+			require.Contains(t, err.Error(), testC.expectedError)
+		})
+	}
+
+	// Success testcases
+	t.Run("when_the_token_is_valid,_it_MUST_return_the_token_subject_and_its_associated_scopes", func(t *testing.T) {
+		scopes := "offline_access read write delete"
+		oidc, requestContext, err := quickConfigSetup(
+			"kid_1",
+			"kid_1",
+			"right_issuer",
+			"right_audience",
+			jwt.MapClaims{
+				"iss":   "right_issuer",
+				"aud":   "right_audience",
+				"sub":   "openfga client",
+				"scope": scopes,
+			},
+			nil,
+		)
+		if err != nil {
+			t.Fatal(err)
+		}
+		authClaims, err := oidc.Authenticate(requestContext)
+		require.Nil(t, err)
+		require.Equal(t, "openfga client", authClaims.Subject)
+		scopesList := strings.Split(scopes, " ")
+		require.Equal(t, len(scopesList), len(authClaims.Scopes))
+		for _, scope := range scopesList {
+			_, ok := authClaims.Scopes[scope]
+			require.True(t, ok)
+		}
+
+	})
+}
+
+// quickConfigSetup sets up a basic configuration for testing purposes.
+func quickConfigSetup(jwkKid, jwtKid, issuerURL, audience string, jwtClaims jwt.MapClaims, privateKeyOverride *rsa.PrivateKey) (*RemoteOidcAuthenticator, context.Context, error) {
+	// Generate JWT signature keys
+	privateKey, publicKey := generateJWTSignatureKeys()
+	if privateKeyOverride != nil {
+		privateKey = privateKeyOverride
+	}
+	// assign mocked JWKS fetching function to global function
```

Comment:
yikes we are using a global function 😱 

are sure that by virtue of having this test we are not overriding the production code? can you do a sanity check and post the result in the PR description?

1. run a local JWK server (a basic server that responds to `.well-known/openid-configuration` with a static file)
2. run `./openfga run --playground-enabled=false --authn-method=oidc --authn-oidc-audience=local --authn-oidc-issuer="http://LOCAL-ADDRESS:LOCAL-PORT"`
4. make a call to OpenFGA's Check API with a token that has a token signed with a different key, you should get "invalid bearer token"

Reply:
@miparnisari could you please clarify what exactly you want to validate with sanity check?

Reply:
i want to be 100% sure that by doing `fetchJWKs = fetchJWK` we are not overriding the production code with test code

Reply:
If this is the purpose, I believe that a sanity test that only checks whether a token is valid is more than enough. 

Given that the tests solely simulate the `fetchJWK` behavior, and as per its design, this function exclusively retrieves JWKs without conducting any validation, I perceive no potential for altering the validation behavior unless there's an oversight on my part.           

Here is the sanity test that performs token validation: 

`make build && ./openfga run --playground-enabled=false  --authn-method=oidc --authn-oidc-issuer="http://localhost:8185/realms/internal"  --authn-oidc-audience="account"`
![image](https://github.com/openfga/openfga/assets/58696746/199cce44-7458-442c-9eb6-01601b81cb93)

Request with valid token:
![image](https://github.com/openfga/openfga/assets/58696746/e13ad937-af91-4dd9-a0bc-a24372b2fdc9)

Request with invalid token: 
![image](https://github.com/openfga/openfga/assets/58696746/c9a188a2-7442-486a-b42a-4dbefbc46a1c)